Replace delete_if with compact

siaw23 opened the pull request:
### Summary

Here's a benchmark showing #compact is about 2.8x faster

```ruby
require 'benchmark/ips'

hash = {
    'a' => 0,
    'b' => '',
    'c' => '',
    'd' => '',
    'e' => nil,
    'f' => nil,
    'g' => '',
    'h' => '',
    'i' => '',
    'j' => ''
  }

def delif(h)
  h.dup.delete_if { |_, v| v.nil? }
end

def comp(h)
  h.compact
end

delif_hash = delif(hash)
raise unless hash.keys.length == 10

comp_hash = comp(hash)
raise unless hash.keys.length == 10

raise unless (delif_hash.keys - comp_hash.keys).empty?

Benchmark.ips do |x|
  x.report('delete_if') do
    delif(hash)
  end

  x.report('compact') do
    comp(hash)
  end

  x.compare!
end


# Warming up --------------------------------------
#            delete_if    61.088k i/100ms
#              compact   164.488k i/100ms
# Calculating -------------------------------------
#            delete_if    591.837k (± 5.1%) i/s -      2.993M in   5.070631s
#              compact      1.674M (± 3.1%) i/s -      8.389M in   5.014899s

# Comparison:
#              compact:  1674391.9 i/s
#            delete_if:   591837.2 i/s - 2.83x  (± 0.00) slower

# Benchmark courtesy of Hartley McGuire.
```

### Other Information

<!-- If there's anything else that's important and relevant to your pull
request, mention that information here. This could include
benchmarks, or other information.

If you are updating any of the CHANGELOG files or are asked to update the
CHANGELOG files by reviewers, please add the CHANGELOG entry at the top of the file.

Finally, if your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

Thanks for contributing to Rails! -->

<!--
Note: Please avoid making *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.
Create a pull request when it is ready for review and feedback
from the Rails team :).
-->

eileencodes commented:
Hi @siaw23 thanks for the PR. Your commit and PR don't mention why you want to make this change. In Rails we don't accept cosmetic changes so this needs to either improve performance or fix a bug for us to decide to merge. Can you help me understand why this change is necessary? Thanks!

skipkayhil commented:
```ruby
require 'benchmark/ips'

Benchmark.ips do |x|
  hash = {
    'a' => 0,
    'b' => '',
    'c' => '',
    'd' => '',
    'e' => nil,
    'f' => nil,
    'g' => '',
    'h' => '',
    'i' => '',
    'j' => ''
  }
  x.report('delete_if') do
    hash.delete_if { |_, v| v.nil? }
  end

  x.report('compact') do
    hash.compact
  end

  x.compare!
end
```

I'm not sure this benchmark is quite right. Since `delete_if` is a destructive operation, `hash` won't have any `nil` elements after the first time the `delete_if` is run. You need to make sure the hash looks the same in each iteration of the benchmark.

(In an unrelated note, it may be worth trying out `compact!` as well since that would be more similar to `delete_if`)

siaw23 commented:
Thanks for the info @eileencodes, my intention was a performance boost, allow me to figure out if this makes sense at all :)

@skipkayhil the problem with `compact!` is that it returns `nil` if no entries were removed. Do you maybe have a better way to benchmark this? 🤔

skipkayhil commented:
> Do you maybe have a better way to benchmark this?

```ruby
require 'benchmark/ips'

hash = {
    'a' => 0,
    'b' => '',
    'c' => '',
    'd' => '',
    'e' => nil,
    'f' => nil,
    'g' => '',
    'h' => '',
    'i' => '',
    'j' => ''
  }

def delif(h)
  h.dup.delete_if { |_, v| v.nil? }
end

def comp(h)
  h.compact
end

delif_hash = delif(hash)
raise unless hash.keys.length == 10

comp_hash = delif(hash)
raise unless hash.keys.length == 10

raise unless (delif_hash.keys - comp_hash.keys).empty?

Benchmark.ips do |x|
  x.report('delete_if') do
    delif(hash)
  end

  x.report('compact') do
    comp(hash)
  end

  x.compare!
end
```

How about this? I added some inline assertions to make sure the logic tracked.

> the problem with `compact!` is that it returns `nil` if no entries were removed

I thought about this some more and realized since we're calling `dup` with `delete_if`, `compact` without `dup` would actually make more sense than `compact!`

skipkayhil commented on actionpack/lib/action_dispatch/request/session.rb:
```diff
@@ -163,7 +163,7 @@ def clear
       def to_hash
         load_for_read!
-        @delegate.dup.delete_if { |_, v| v.nil? }
+        @delegate.dup.compact
```
```suggestion
        @delegate.compact
```

since `compact` is returning a new hash anyway, we shouldn't need to `dup` as well

Commit:
Replace delete_if with compact for performance

Here's a benchmark showing #compact is about 2.8x faster

# require 'benchmark/ips'

# hash = {
#     'a' => 0,
#     'b' => '',
#     'c' => '',
#     'd' => '',
#     'e' => nil,
#     'f' => nil,
#     'g' => '',
#     'h' => '',
#     'i' => '',
#     'j' => ''
#   }

# def delif(h)
#   h.dup.delete_if { |_, v| v.nil? }
# end

# def comp(h)
#   h.compact
# end

# delif_hash = delif(hash)
# raise unless hash.keys.length == 10

# comp_hash = comp(hash)
# raise unless hash.keys.length == 10

# raise unless (delif_hash.keys - comp_hash.keys).empty?

# Benchmark.ips do |x|
#   x.report('delete_if') do
#     delif(hash)
#   end

#   x.report('compact') do
#     comp(hash)
#   end

#   x.compare!
# end

# Warming up --------------------------------------
#            delete_if    61.088k i/100ms
#              compact   164.488k i/100ms
# Calculating -------------------------------------
#            delete_if    591.837k (± 5.1%) i/s -      2.993M in   5.070631s
#              compact      1.674M (± 3.1%) i/s -      8.389M in   5.014899s

# Comparison:
#              compact:  1674391.9 i/s
#            delete_if:   591837.2 i/s - 2.83x  (± 0.00) slower

# Benchmark courtesy of Hartley McGuire.

siaw23 commented:
@eileencodes @skipkayhil are we happy with the changes? :)
